### PR TITLE
[spmd_types] spmd.local region in ChunkedCELoss

### PR DESCRIPTION
**Pull request by pianpwk**
No immediate effect, just prepping for next commit.

I decided not to use `spmd.local_map` mainly because it needs refactoring out the chunked loss body as standalone function, and the assert_type/assert_type_like following the region input is more convenient, than specifying the full placements for both loss & hidden states..

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3690
* #3673
* #3655
* #3664
* #3654
* #3278
* #3694
* #3472
* __->__ #3632
* #3653

### CODE REVIEW

**Comment by fegin** [comment]
Can you add the reason why this is required to make spmd_types work? Want to understand the issue.

**Comment by pianpwk** [comment]
> Can you add the reason why this is required to make spmd_types work? Want to understand the issue.

it's really not, gonna update this

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -556,63 +556,64 @@ def _chunk(t):
...
+                    lm_head.set_requires_gradient_sync(  # pyrefly: ignore[not-callable]
+                        True, recurse=False
+                    )
+
+                logits = lm_head(h_chunk)
+
+                chunk_loss = self.fn(logits, label_chunk)
+                if global_valid_tokens is not None:
+                    chunk_loss = chunk_loss / global_valid_tokens
+                total_loss = total_loss + chunk_loss.detach()
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:588]: What are the local spmd_types for each operands?

It is a bit surprising to see this region working with local type checking.

Comment (pianpwk) [reply torchtitan/components/loss.py:588]: it needs the next commit changes, this PR is a no-op, just prep for reviewability

total_loss needs mutate_type to be init as P@DP/CP, I@TP

chunk_loss exits local SPMD region as V@DP/CP, I@TP, but gets reinterpreted to P@DP/CP (lose partial inference ability in local)

global_valid_tokens we changed to float

**Review by tianyu-l — APPROVED** [review]
